File: go/2022/day05/day05.go
package main

import (
	"flag"
	"fmt"
	"os"
	"strings"
)

var inputFile = flag.String("inputFile", "input.txt", "Relative file path to use as input")

func main() {
	flag.Parse()
	bytes, err := os.ReadFile(*inputFile)
	if err != nil {
		panic("Unable to read the input file")
	}

	input := string(bytes)

	fmt.Printf("AoC 2022, Day5, Part1 solution is: %s \n", part1(input))
	fmt.Printf("AoC 2022, Day5, Part2 solution is: %s \n", part2(input))
}

func part1(input string) string {
	stacks, instructions := parseInput(input)

	for _, instruction := range instructions {
		instruction.moveReversed(stacks)
	}

	return getResponse(stacks)
}

func part2(input string) string {
	stacks, instructions := parseInput(input)
	for _, instruction := range instructions {
		instruction.moveInTheSameOrder(stacks)
	}
	return getResponse(stacks)
}

type instruction struct {
	nrCrates int
	from     int
	to       int
}

func (instr *instruction) moveReversed(stack [][]string) {
	fromStack := stack[instr.from]
	toStack := stack[instr.to]

	movingCrates := fromStack[len(fromStack)-instr.nrCrates:]
	stack[instr.from] = fromStack[:len(fromStack)-instr.nrCrates]

	for i := len(movingCrates) - 1; i >= 0; i-- {
		toStack = append(toStack, movingCrates[i])
	}
	stack[instr.to] = toStack
}

func (instr *instruction) moveInTheSameOrder(stack [][]string) {
	fromStack := stack[instr.from]
	toStack := stack[instr.to]

	movingCrates := fromStack[len(fromStack)-instr.nrCrates:]
	stack[instr.from] = fromStack[:len(fromStack)-instr.nrCrates]

	toStack = append(toStack, movingCrates...)
	stack[instr.to] = toStack
}

func parseInput(input string) ([][]string, []instruction) {
	parts := strings.Split(input, "\n\n")
	rawStacks, rawRules := parts[0], parts[1]
	return getStacks(rawStacks), getInstructions(rawRules)
}

func getInstructions(rules string) []instruction {
	instructions := make([]instruction, 0)

	for _, line := range strings.Split(rules, "\n") {
		instruction := instruction{}
		fmt.Sscanf(line, "move %d from %d to %d", &instruction.nrCrates, &instruction.from, &instruction.to)
		instruction.from--
		instruction.to--
		instructions = append(instructions, instruction)
	}

	return instructions
}

func getStacks(rawStacks string) [][]string {
	lines := strings.Split(rawStacks, "\n")

	nrOfStacks := 0
	for _, n := range strings.Split(lines[len(lines)-1], " ") {
		if n != "" {
			nrOfStacks++
		}
	}

	stacks := [][]string{}

	for i := 1; i < len(lines[0]); i += 4 {
		stack := make([]string, 0)

		for j := len(lines) - 2; j >= 0; j-- {
			currentLine := lines[j]
			if currentLine[i] != ' ' {
				stack = append(stack, string(currentLine[i]))
			}
		}
		stacks = append(stacks, stack)
	}
	return stacks
}

func getResponse(stacks [][]string) string {
	sb := strings.Builder{}
	for _, stack := range stacks {
		sb.WriteString(stack[len(stack)-1])
	}
	return sb.String()
}

File: python/2022/day05/day05.py
from dataclasses import dataclass
from typing import List


@dataclass
class Action:
    nr_of_crates: int
    from_stack: int
    to_stack: int


def create_actions(input: str) -> List[Action]:
    actions = []
    for line in input.split("\n"):
        parts = line.split(" ")
        actions.append(
            Action(
                nr_of_crates=int(parts[1]),
                from_stack=int(parts[3]),
                to_stack=int(parts[5]),
            )
        )
    return actions


def create_stacks(input: str) -> List[List[str]]:
    lines = input.split("\n")
    nr_of_stacks = [int(x) for x in lines[-1].split(" ") if x != ""]
    stacks = [[] for _ in range(len(nr_of_stacks))]

    for i in range(len(lines) - 1):
        current_line = lines[i]
        for j in range(1, len(current_line), +4):
            stack_index = j // 4
            if current_line[j] != " ":
                stacks[stack_index].append(current_line[j])

    return stacks


def run(input: str, reverse: bool) -> str:
    stacks = create_stacks(parse_input(input)[0])
    actions = create_actions(parse_input(input)[1])

    for action in actions:
        to_stack = action.to_stack
        from_stack = action.from_stack
        nr_of_crates = action.nr_of_crates

        creates = stacks[from_stack - 1][:nr_of_crates]
        if reverse:
            creates.reverse()

        stacks[to_stack - 1] = creates + stacks[to_stack - 1]
        stacks[from_stack - 1] = stacks[from_stack - 1][action.nr_of_crates :]

    first_crates = [stack[0] for stack in stacks]
    return "".join(first_crates)


def parse_input(input: str) -> List[str]:
    with open(input) as fn:
        raw_data = fn.read()
        parts = raw_data.split("\n\n")
    return parts


def main():
    puzzle_input_path = '2022/day05/input.txt'
    print(f"AoC2022 Day5, Part1 solution is: {run(input=puzzle_input_path, reverse = True)}")
    print(f"AoC2022 Day5, Part2 solution is: {run(input=puzzle_input_path, reverse = False)}")


if __name__ == "__main__":
    main()
